src/neuron/container/soa_container.hpp

```diff
@@ -0,0 +1,1123 @@
+ * data values and more generic handle types such as @ref
+ * neuron::container::data_handle<T> and @ref
+ * neuron::container::generic_data_handle.
+ */
+template <typename Storage, typename... Tags>
+struct soa {
+    /**
+     * @brief Construct with default-constructed tag type instances.
+     */
+    soa()
+        : soa(Tags{}...) {}
+
+    /**
+     * @brief Construct with specific tag instances.
+     *
+     * This is useful if the tag types are not empty, for example if the number
+     * of times a column is duplicated is a runtime value.
+     */
+    soa(Tags... tag_instances)
+        : m_data{std::move(tag_instances)...} {}
+
+    /**
+     * @brief @ref soa is not movable
+     *
+     * This is to make it harder to accidentally invalidate pointers-to-storage
+     * in handles.
+     */
+    soa(soa&&) = delete;
+
+    /**
+     * @brief @ref soa is not copiable
+     *
+     * This is partly to make it harder to accidentally invalidate
+     * pointers-to-storage in handles, and partly because it could be very
+     * expensive so it might be better to be more explicit.
+     */
+    soa(soa const&) = delete;
+
+    /**
+     * @brief @ref soa is not move assignable
+     *
+     * For the same reason it isn't movable.
+     */
+    soa& operator=(soa&&) = delete;
+
+    /**
+     * @brief @ref soa is not copy assignable
+     *
+     * For the same reasons it isn't copy constructible
+     */
+    soa& operator=(soa const&) = delete;
+
+    /**
+     * @brief Get the size of the container.
+     */
+    [[nodiscard]] std::size_t size() const {
+        // Check our various std::vector members are still the same size as each
+        // other. This check could be omitted in release builds...
+        auto const check_size = m_indices.size();
+        for_all_vectors(
+            [check_size](auto const& tag, auto const& vec, int field_index, int array_dim) {
+                auto const size = vec.size();
+                assert(size % array_dim == 0);
+                assert(size / array_dim == check_size);
+            });
+        return check_size;
+    }
+
+    /**
+     * @brief Test if the container is empty.
+     */
+    [[nodiscard]] bool empty() const {
+        auto const result = m_indices.empty();
+        for_all_vectors([result](auto const& tag, auto const& vec, int field_index, int array_dim) {
+            assert(vec.empty() == result);
+        });
+        return result;
+    }
+
+  private:
+    /**
+     * @brief Remove the @f$i^{\text{th}}@f$ row from the container.
+     *
+     * This is currently implemented by swapping the last element into position
+     * @f$i@f$ (if those are not the same element) and reducing the size by one.
+     * Iterators to the last element and the deleted element will be
+     * invalidated.
+     */
+    void erase(std::size_t i) {
+        if (m_frozen_count) {
+            throw_error("erase() called on a frozen structure");
```

Comment:
It should be reachable via https://github.com/neuronsimulator/nrn/blob/08d4665ef4dfbd7ee945cfcb3744bdd0ffbbbb39/src/neuron/container/soa_identifier.hpp#L138, but it's awkward because that triggers `std::terminate`. I've tried to add a test for this, but it may need some tuning to make sure that the coverage data is flushed to disk even when we call `std::terminate` (e.g. https://en.cppreference.com/w/cpp/error/set_terminate, https://stackoverflow.com/questions/68550611/how-to-get-gcov-gcda-coverage-if-test-run-was-terminated).